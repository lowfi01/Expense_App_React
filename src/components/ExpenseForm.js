import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

// // use moment
// const now = moment().format('MMMM Do, YYYY, h:mm:ss a');
// console.log(now);


class ExpenseFrom extends React.Component{
  constructor(props) {
    super(props);

    this.state = {
        description: props.expense ? props.expense.description : '',
        note: props.expense ? props.expense.note : '',
        amount: props.expense ? (props.expense.amount / 100).toString() : '',
        createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
        calenderFocused: false,
        error: ''
    }
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({
      description
    }));
  }

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({
      note
    }));
  }

  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({
        amount
      }))
    }
  }

  // gets created with the moment object
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState( () => ({createdAt}));
    }

  }

  onFocuseChange = ({ focused }) => {
    this.setState(() => ({
      calenderFocused: focused
    }))
  };

  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: 'Please provide description and amount'
      }))
    } else {
      this.setState(() => ({
        error: ''
      }))

      // Add validation layer, before sending state object to the parent
      this.props.onSubmit({
        description: this.state.description,
        note: this.state.note,
        amount:  parseFloat(this.state.amount, 10) * 100,
        createdAt:  this.state.createdAt.valueOf(),
      });
    }

  }

  render() {
    return (
      <div>
        { this.state.error && <p>{this.state.error}</p> }
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calenderFocused}
            onFocusChange={this.onFocuseChange}
            numberOfMonths={1}
            isOutsideRange={(day) => false}
          />
          <textarea
            placeholder="Add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
            />
          <button type="submit" >Submit</button>
        </form>
      </div>
    )
  }
}

export default ExpenseFrom;