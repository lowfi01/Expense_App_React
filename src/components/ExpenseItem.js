import React from 'react';
import { Link } from 'react-router-dom';

export const ExpenseItem = ({id, description, amount, createdAt, dispatch}) => (
    <div>
      <Link to={`/edit/${id}`}>
        <h3>{description}</h3>
      </Link>
      <p>{amount}-{createdAt}</p>

    </div>
  )

export default ExpenseItem;