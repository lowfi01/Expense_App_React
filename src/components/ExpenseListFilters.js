import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';

// Import Action Generator
import { setStartDate, setEndDate, setTextFilter, sortByAmount, sortByDate } from '../actions/filters'

class ExpenseListFilters extends React.Component {
  state = {
    calenderFocused: null
  }

  onFocusedChanged = (calenderFocused) => {
      this.setState(() => ({
        calenderFocused
      }));
  }

  onDatesChange = ({startDate, endDate}) => {
    this.props.dispatch(setStartDate(startDate));
    this.props.dispatch(setEndDate(endDate));
  }

  render() {
    return (
      <div>
      <input
      type="text"
      value={this.props.filters.text}  // value is defined by state value
      onChange={(e) => {
        this.props.dispatch(setTextFilter(e.target.value));
      }}/>

      <select
        value={this.props.filters.sortBy}
        onChange={(e) => {
         if(e.target.value === 'date') {
          this.props.dispatch(sortByDate());
         }
         if (e.target.value === 'amount') {
          this.props.dispatch(sortByAmount());
         }

        }}>
        <option value="date">date</option>
        <option value="amount">amount</option>
      </select>

      <DateRangePicker
        startDate={this.props.filters.startDate}
        endDate={this.props.filters.endDate}
        onDatesChange={this.onDatesChange}
        focusedInput={this.state.calenderFocused}
        onFocusChange={this.onFocusedChanged}
        numberOfMonths={1}
        isOutsideRange={() => false}
        showClearDates={true}
      />
    </div>
    )
  }


}

const mapStateToProps = (state) => ({
    filters: state.filters
  })


export default connect(mapStateToProps)(ExpenseListFilters);