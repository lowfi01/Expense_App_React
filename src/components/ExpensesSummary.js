import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';

import SelectedExpenses from '../selectors/expenses';
import ExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = ({expenseCount, expensesTotal}) => {
  const expenseWord = expenseCount === 0 ? 'expenses' : 'expense';
  const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00');
  return (
    <div>
    <h1> Viewing {expenseCount} {expenseWord} with the value of {formattedExpensesTotal} </h1>
    </div>
  )
}

const mapStateToProps = (state) => ({
  expensesCount: SelectedExpenses(state.expenses, state.filters).length,
  expensesTotal: ExpensesTotal(SelectedExpenses(state.expenses, state.filters))
  })

export default connect(mapStateToProps)(ExpensesSummary);
