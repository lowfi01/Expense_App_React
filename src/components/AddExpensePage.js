import React from 'react';
import {connect} from 'react-redux';
import ReactGA from 'react-ga';
import ExpenseForm from './ExpenseForm';

// import { addExpense } from '../actions/expenses';
import { startAddExpense } from '../actions/expenses';

const AddExpensePage = (props) => {
  const onSubmit = (expense) => {
      ReactGA.event({
        category: 'Create Expense',
        action: 'click button',
        label: 'expenses'
      });
      props.dispatch(startAddExpense(expense));
      props.history.push('/');
    }
  return(
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Expense</h1>
          </div>
        </div>

        <div className="content-container">
          <ExpenseForm onSubmit={onSubmit} />
        </div>

      </div>
    )
  }

export default connect()(AddExpensePage);