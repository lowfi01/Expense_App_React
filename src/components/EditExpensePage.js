import React from 'react';
import { connect } from 'react-redux';
import ReactGA from 'react-ga';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

const EditExpensePage = (props) => {
  console.log('props');
  return (
    <div>
      <div className="page-header">
        <div className="content-container">
          <h1 className="page-header__title">Edit Expense</h1>
        </div>
      </div>

      <div className="content-container">
        <ExpenseForm
          expense={props.expense}
          onSubmit={(expense) => {
              ReactGA.event({
                category: 'Edit Expense',
                action: 'click button',
                label: 'expenses'
              });
              props.dispatch(startEditExpense(props.match.params.id, expense));
              props.history.push('/');

          }}
        />
        <button
          className="button button--secondary"
          onClick={(e) => {
          e.preventDefault();
          ReactGA.event({
            category: 'Remove Expense',
            action: 'click button',
            label: 'expenses'
          });
          props.dispatch(startRemoveExpense({ id : props.expense.id}));
          props.history.push('/');

        }}
        >
          Remove Expense
        </button>
      </div>
    </div>
  );
};

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  })

export default connect(mapStateToProps)(EditExpensePage);
