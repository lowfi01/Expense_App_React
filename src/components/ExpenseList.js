import React from 'react';
import { connect } from 'react-redux';

import selectExpenses from '../selectors/expenses';


import ExpenseItem from './ExpenseItem';

// we export this component as a named component for testing
export const ExpenseList = (props) => (
    <div>

      {
        props.expenses.length === 0 ? (
          <p>No expenses</p>
        ) : (
          props.expenses.map((expense) =>
          <ExpenseItem key={expense.id} {...expense}/>)
        )
      }

    </div>
  );


const mapStateToProps = (state) => ({
  // selectExpenses is our filter function from selectors
  expenses: selectExpenses(state.expenses, state.filters)
 }
)

export default connect(mapStateToProps)(ExpenseList);


// connect is a HOC - Higher Order component, takes argument of component then return component.
// Note - connect also has an option to pass agruments
// We pass the information we want to connect, these are functions.
// The function gains access to the state from the argument


// const ConnectedExpenseList = connect((state) => ({
//     expenses: state.expenses
//    }
//   ))(ExpenseList);

// export default ConnectedExpenseList