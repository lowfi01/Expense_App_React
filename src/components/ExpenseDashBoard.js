import React from 'react';

import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary';

export const ExpenseDashBoard = (props) => (
    <div>
      <h1>Expense DashBoard</h1>
      <ExpenseListFilters />
      <ExpenseList />
      <ExpensesSummary />
    </div>

  )

export default ExpenseDashBoard;