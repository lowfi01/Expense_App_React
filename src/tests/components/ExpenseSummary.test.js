import React from 'react';
import { shallow } from 'enzyme';
import {ExpensesSummary} from '../../components/ExpensesSummary';

test('should correctly render ExpensesSummary report with 1 expenses', () => {
  const wrapper = shallow(<ExpensesSummary expensesCount={1} expensesTotal={200}/>);
  expect(wrapper).toMatchSnapshot();
})


test('should correctly render ExpensesSummary report with multiple expenses', () => {
  const wrapper = shallow(<ExpensesSummary expensesCount={4} expensesTotal={24999}/>);
  expect(wrapper).toMatchSnapshot();
})


test('should correctly render Expenses Summary report with ', () => {
  const wrapper = shallow(<ExpensesSummary expensesCount={0} expensesTotal={0}/>);
  expect(wrapper).toMatchSnapshot();
})
