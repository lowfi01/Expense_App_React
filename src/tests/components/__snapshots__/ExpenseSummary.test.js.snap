// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render Expenses Summary report with  1`] = `
<div>
  <h1>
     Viewing 
     
    expense
     with the value of 
    $0.00
     
  </h1>
</div>
`;

exports[`should correctly render ExpensesSummary report with 1 expenses 1`] = `
<div>
  <h1>
     Viewing 
     
    expense
     with the value of 
    $2.00
     
  </h1>
</div>
`;

exports[`should correctly render ExpensesSummary report with multiple expenses 1`] = `
<div>
  <h1>
     Viewing 
     
    expense
     with the value of 
    $249.99
     
  </h1>
</div>
`;
